src/model/document/document.ts

```diff
@@ -5,12 +5,25 @@ import { Vector2 } from '../vectors'
 import { readonly, shallowReactive, shallowRef, ref, watch } from 'vue'
 import arrayUtils from '../array-utils'
 import { ScopeElement, ScopeElementType } from './elements/scope-element'
+import { ExpressionElement, ExpressionElementType } from './elements/expression-element'
+
+type JsonType = null | boolean | number | string | JsonType[] | Vector2 | { [prop: string]: JsonType }
+
+type SerializedDataType = {
+  expressionElements: JsonType[]
+  scopeElement: JsonType[]
+}
```

Comment:
Glad to see you writing your own types!
If you want to keep this very general/generic, you could have a single `elements: JsonType[]` list. And then, every element has to specify its type. 

Then, further down you could have something along the lines of
```js
let elementType = elementTypes[elements[i].type]
elementType.deserializeElement(elements[i])
```

Reply:
I was trying to think of a way to do just this! Now I don't need a bit of code for each element type.